File: 4007-maximum-total-subarray-value-ii/maximum-total-subarray-value-ii.java
class Solution {
    public long maxTotalValue(int[] nums, int k) {
        int n = nums.length;
        if (n == 0 || k == 0) return 0;

        int log = (int) (Math.log(n) / Math.log(2)) + 1;
        int[][] stMax = new int[n][log];
        int[][] stMin = new int[n][log];

        for (int i = 0; i < n; i++) {
            stMax[i][0] = nums[i];
            stMin[i][0] = nums[i];
        }

        for (int j = 1; j < log; j++) {
            int length = 1 << (j - 1);
            for (int i = 0; i + (1 << j) <= n; i++) {
                stMax[i][j] = Math.max(stMax[i][j - 1], stMax[i + length][j - 1]);
                stMin[i][j] = Math.min(stMin[i][j - 1], stMin[i + length][j - 1]);
            }
        }

        int[] lg = new int[n + 1];
        for (int i = 2; i <= n; i++) {
            lg[i] = lg[i / 2] + 1;
        } 
        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));

        for (int l = 0; l < n; l++) {
            long val = queryVal(stMax, stMin, lg, l, n - 1);
            pq.offer(new long[]{-val, l, n - 1});
        }

        long totalVal = 0;
        for (int i = 0; i < k; i++) {
            long[] curr = pq.poll();
            long val = -curr[0];
            int l = (int) curr[1];
            int r = (int) curr[2];

            totalVal += val;

            if (r > l) {
                long nextVal = queryVal(stMax, stMin, lg, l, r - 1);
                pq.offer(new long[]{-nextVal, l, r - 1});
            }
        }

        return totalVal;
    }

    private long queryVal(int[][] stMax, int[][] stMin, int[] lg, int l, int r) {
        int length = r - l + 1;
        int j = lg[length];
        int mx = Math.max(stMax[l][j], stMax[r - (1 << j) + 1][j]);
        int mn = Math.min(stMin[l][j], stMin[r - (1 << j) + 1][j]);
        return (long) mx - mn;
    }
}

File: 4007-maximum-total-subarray-value-ii/maximum-total-subarray-value-ii.py
class Solution:
    def maxTotalValue(self, nums: List[int], k: int) -> int:
        n = len(nums)
        if n == 0 or k == 0:
            return 0 
        LOG = math.floor(math.log2(n)) + 1
        st_max = [[0] * LOG for _ in range(n)]
        st_min = [[0] * LOG for _ in range(n)]
        
        for i in range(n):
            st_max[i][0] = nums[i]
            st_min[i][0] = nums[i]
            
        for j in range(1, LOG):
            length = 1 << (j - 1)
            for i in range(n - (1 << j) + 1):
                st_max[i][j] = max(st_max[i][j - 1], st_max[i + length][j - 1])
                st_min[i][j] = min(st_min[i][j - 1], st_min[i + length][j - 1])
                
        lg = [0] * (n + 1)
        for i in range(2, n + 1):
            lg[i] = lg[i // 2] + 1

        def query_val(l: int, r: int) -> int:
            length = r - l + 1
            j = lg[length]
            mx = max(st_max[l][j], st_max[r - (1 << j) + 1][j])
            mn = min(st_min[l][j], st_min[r - (1 << j) + 1][j])
            return mx - mn 
        heap = []
        for l in range(n):
            val = query_val(l, n - 1)
            heapq.heappush(heap, (-val, l, n - 1))
            
        total_val = 0
        for _ in range(k):
            neg_val, l, r = heapq.heappop(heap)
            total_val += -neg_val
            
            if r > l:
                next_val = query_val(l, r - 1)
                heapq.heappush(heap, (-next_val, l, r - 1))
                
        return total_val